### PR TITLE
r/aws_lb_listener: add support for additional SSL certificates

### PR DESCRIPTION
**Pull request by oarmstrong**
This PR is for #1853.

We add an additional argument (`additional_certificate_arns`) to the `aws_lb_listener` resource. This argument allows for adding additional non-default certificates to a listener.

I'm well aware that there are currently no tests or documentation and that still needs to be done. It has been manually tested with as many cases as I could think of though. The existing test suite still runs, but I have not run the acceptance tests at all.

If at all possible, some feedback on the general style of my code would be great. Variable naming and error checking are two points that I've been particularly bad at in the past. I'm also not sure about defining https://github.com/terraform-providers/terraform-provider-aws/pull/2203/files#diff-6f5ff88b797d88c79786bcdb04a23eefR314 in line like that. Do we have an existing function somewhere that does the same job or should I move this elsewhere?

Another suggestion in the linked issue is for it to be possible to define additional certificates as a separate resource - must like security group rules. I'm not too sure if this design would allow for that?

As a final question, how would you like the commits before this can be merged in? Squished? Some kind of prefix to the messages?

Appreciate your time. Hopefully there isn't anything too major wrong with my code and I can get the documentation and tests done quickly too.

### CODE REVIEW

**Comment by oarmstrong** [comment]
Documentation written. I can also confirm that the acceptance suit for this resource passes. Currently writing the updated tests for this argument now.

**Comment by oarmstrong** [comment]
That's just added the tests.

I've also modified the corresponding data source, not sure if you'd prefer that in a separate PR though.

**Comment by oarmstrong** [comment]
It may make more sense to rename the original `certificate_arn` argument to `default_certificate_arn` as the API and documentation seems to refer to it as that. Not sure how you'd feel about the breaking change though.

**Comment by ozbillwang** [comment]
@oarmstrong 

Thanks to raise this PR to add this feature. When will it be merged?

Regarding argument naming, I am fine for  `additional_certificate_arns` more than to rename `certificate_arn` to `default_certificate_arn` to keep version compatible.

**Comment by oarmstrong** [comment]
> When will it be merged?

It's good to go from me. Just awaiting feedback from the maintainers.

**Comment by ozbillwang** [comment]
Seems we missed the release in 0.11, hope to get it merged in next release.

**Comment by Puneeth-n** [comment]
@ozbillwang Once, the PR is merged, the maintainers update the CHANGELOG

**Comment by tatsuo48** [comment]
Is there anyone who knows when to merge?
I'm waiting this merge 😢

**Comment by oarmstrong** [comment]
I'm still not 100% with it and assume it will need some polish. Just want a bit of a review at this point.

**Comment by Puneeth-n** [comment]
@radeksimko @Ninir any idea when this will be merged :)

**Comment by oarmstrong** [comment]
I'm able to get on any review points pretty much immediately!

File: aws/resource_aws_lb_listener.go [hunk 1]
```diff
@@ -147,7 +154,38 @@ func resourceAwsLbListenerCreate(d *schema.ResourceData, meta interface{}) error
...
+		if carns, ok := d.GetOk("additional_certificate_arns"); ok {
+			for _, carn := range expandStringList(carns.(*schema.Set).List()) {
+				_, err := elbconn.AddListenerCertificates(&elbv2.AddListenerCertificatesInput{
+					Certificates: []*elbv2.Certificate{&elbv2.Certificate{
+						CertificateArn: carn,
+					}},
+					ListenerArn: aws.String(larn),
+				})
+				if awsErr, ok := err.(awserr.Error); ok {
+					if awsErr.Code() == "TooManyCertificates" || awsErr.Code() == "CertificateNotFound" {
```
Comment (radeksimko) [inline aws/resource_aws_lb_listener.go]: What is the reason for retrying on `TooManyCertificates` here? 🤔

File: aws/resource_aws_lb_listener.go [hunk 2]
```diff
@@ -147,7 +154,38 @@ func resourceAwsLbListenerCreate(d *schema.ResourceData, meta interface{}) error
...
+					return resource.NonRetryableError(err)
+				}
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return errwrap.Wrapf("Error adding certificates to LB Listener: {{err}}", err)
```
Comment (radeksimko) [inline aws/resource_aws_lb_listener.go]: `errwrap` is typically useful in situations where we need to keep track of the error type. In this case error will only ever be escalated to the UI where all that matters is strings ("stringifyed" error message), not the original type.

In other words `fmt.Errorf()` will do the job, IMO.

File: aws/resource_aws_lb_listener.go [hunk 3]
```diff
@@ -183,6 +221,23 @@ func resourceAwsLbListenerRead(d *schema.ResourceData, meta interface{}) error {
 		d.Set("certificate_arn", listener.Certificates[0].CertificateArn)
 	}
 
+	certs, err := elbconn.DescribeListenerCertificates(&elbv2.DescribeListenerCertificatesInput{
+		ListenerArn: aws.String(d.Id()),
+	})
+	if err != nil {
+		return errwrap.Wrapf("Error retrieving ListenerCertficates: {{err}}", err)
```
Comment (radeksimko) [inline aws/resource_aws_lb_listener.go]: `errwrap` is typically useful in situations where we need to keep track of the error type. In this case error will only ever be escalated to the UI where all that matters is strings ("stringifyed" error message), not the original type.

In other words `fmt.Errorf()` will do the job, IMO.

File: aws/resource_aws_lb_listener.go [hunk 4]
```diff
@@ -236,6 +291,62 @@ func resourceAwsLbListenerUpdate(d *schema.ResourceData, meta interface{}) error
...
+	var carnList []*string
+	if carns, ok := d.GetOk("additional_certificate_arns"); ok {
+		carnList = expandStringList(carns.(*schema.Set).List())
+	}
+
+	awsCerts, err := elbconn.DescribeListenerCertificates(&elbv2.DescribeListenerCertificatesInput{
+		ListenerArn: aws.String(d.Id()),
+	})
+	if err != nil {
+		return errwrap.Wrapf("Error retrieving ListenerCertficates: {{err}}", err)
```
Comment (radeksimko) [inline aws/resource_aws_lb_listener.go]: `errwrap` is typically useful in situations where we need to keep track of the error type. In this case error will only ever be escalated to the UI where all that matters is strings ("stringifyed" error message), not the original type.

In other words `fmt.Errorf()` will do the job, IMO.

File: aws/resource_aws_lb_listener.go [hunk 5]
```diff
@@ -236,6 +291,62 @@ func resourceAwsLbListenerUpdate(d *schema.ResourceData, meta interface{}) error
...
+	for _, carn := range awsCarnList {
+		if !stringInSlice(carn, carnList) {
+			_, err := elbconn.RemoveListenerCertificates(&elbv2.RemoveListenerCertificatesInput{
+				Certificates: []*elbv2.Certificate{&elbv2.Certificate{
+					CertificateArn: carn,
+				}},
+				ListenerArn: aws.String(d.Id()),
+			})
+			if err != nil {
+				return errwrap.Wrapf("Error modifying LB Listener: {{err}}", err)
```
Comment (radeksimko) [inline aws/resource_aws_lb_listener.go]: `errwrap` is typically useful in situations where we need to keep track of the error type. In this case error will only ever be escalated to the UI where all that matters is strings ("stringifyed" error message), not the original type.

In other words `fmt.Errorf()` will do the job, IMO.

File: aws/resource_aws_lb_listener.go [hunk 6]
```diff
@@ -236,6 +291,62 @@ func resourceAwsLbListenerUpdate(d *schema.ResourceData, meta interface{}) error
...
+	for _, carn := range carnList {
+		if !stringInSlice(carn, awsCarnList) {
+			_, err := elbconn.AddListenerCertificates(&elbv2.AddListenerCertificatesInput{
+				Certificates: []*elbv2.Certificate{&elbv2.Certificate{
+					CertificateArn: carn,
+				}},
+				ListenerArn: aws.String(d.Id()),
+			})
+			if err != nil {
+				return errwrap.Wrapf("Error modifying LB Listener: {{err}}", err)
```
Comment (radeksimko) [inline aws/resource_aws_lb_listener.go]: `errwrap` is typically useful in situations where we need to keep track of the error type. In this case error will only ever be escalated to the UI where all that matters is strings ("stringifyed" error message), not the original type.

In other words `fmt.Errorf()` will do the job, IMO.

**Comment by Puneeth-n** [comment]
@radeksimko Great point. I was thinking on similar lines. However with `aws_security_group_rule` how does it behave when I am adding and removing the same rule under different "resource names" ? does terraform error out or does terraform detect that we are adding the same  rule twice under different names?

**Comment by oarmstrong** [comment]
Thanks for the review. I'll get on to it a little later.

File: aws/resource_aws_lb_listener.go (hunk 1 above)
Comment (oarmstrong) [reply aws/resource_aws_lb_listener.go]: Only as I essentially copied `CertificateNotFound`. Neither makes sense to retry on IMO.

File: aws/resource_aws_lb_listener.go (hunk 2 above)
Comment (oarmstrong) [reply aws/resource_aws_lb_listener.go]: Right. I'm with you. Comment noted for all instances.

**Comment by oarmstrong** [comment]
So, the reason for this resource. Basically I was attempting to get this done quickly and simply so editing an existing resource felt much easier for me. It was noted in the issue that a separate resource would be useful too and I agreed 100%. I wasn't aware of how impossible it was to have this resource as well as a separate resource (e.g. the security group case).

It's probably best to do it right from the outset then and just abandon this attempt and go straight for the new resource. If that's your opinion for the "correct" implementation (and I tend to agree with that too).

I can take a crack at the new resource but I may need guidance. Would you prefer to close this PR and for myself to start a new one when I have a first attempt?  I'll reuse a lot of the code from this attempt though as it seems to have passed your initial review.

Thanks for your time!

/cc @radeksimko

**Comment by radeksimko** [comment]
> I can take a crack at the new resource but I may need guidance.

Sure, feel free to ask. Just FYI I will be mostly unavailable from the end of this week until the EOY.

> Would you prefer to close this PR and for myself to start a new one when I have a first attempt? I'll reuse a lot of the code from this attempt though as it seems to have passed your initial review.

Yep, the code looked good - you can reuse most of it and yes it's probably best to close this PR.

If you need some guidance, take a look at any merged `new-resource` PR, e.g. https://github.com/terraform-providers/terraform-provider-aws/pull/2154/files - also note the docs updates.

**Comment by oarmstrong** [comment]
Cheers. I'll close this PR and open a new one with the revised method.

**Pull request closed without merging**